[FIX] web: do not discard falsy values for property fields of type int and float

julien-banken opened the pull request:
Currently, the user can not store the value "0" for the property fields of type integer and the value "0.0" for the property fields of type decimal: Those values get discarded whenever the user unfocuses the input field.

Steps to reproduce:
1. Create a property field of type integer
2. Type "0" in the input field
3. Unfocus the input field

=> The value "0" get discarded

This commit will ensure that those values do not get discarded when the user unfocuses the input field and that they will properly be saved in the database.

task-3226202
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/115030).

caburj commented on addons/web/static/src/views/fields/properties/property_value.js:
```diff
@@ -77,9 +77,11 @@ export class PropertyValue extends Component {
 
-        if (this.props.type === "float") {
+        if (this.props.type === "integer") {
+            return value || 0;
+        } else if (this.props.type === "float") {
             // force to show at least 1 digit, even for integers
-            return value;
+            return value || 0.0;
```
I believe you can do something like this because 0.0 and 0 are just equal.
```js
if (["float", "integer"].includes(this.props.type)) {
  return value || 0;
}
```

caburj commented on addons/web/static/tests/views/fields/properties_field_tests.js:
```diff
@@ -561,10 +561,10 @@ QUnit.module("Fields", (hooks) => {
 
         await editValue("2", "2");
-        await editValue("2.11", "");
-        await editValue("azerty", "", "Wrong integer value should be interpreted as empty");
+        await editValue("2.11", "0");
+        await editValue("azerty", "0", "Wrong integer value should be interpreted as empty");
         await editValue("1,2,3,4,5,6", "123456");
-        await editValue("1,2,3,4,5,6.1,2,3", "");
+        await editValue("1,2,3,4,5,6.1,2,3", "0");
```
Not really sure about this. Why would you fallback to 0 if the input is non-numeric? The test to be added should be:
```js
await editValue("0", "0");
await editValue("0.0", "0");
```

caburj replied:
I'm thinking of just keeping the current behavior. Basically, we should just focus on "zero as input".

caburj commented on addons/web/static/src/views/fields/properties/property_value.js:
```diff
@@ -149,6 +149,10 @@ export class PropertyValue extends Component {
 
         if (this.props.type === "many2one" && value && value.length === 2) {
             return formatMany2one(value);
+        } else if (this.props.type === "integer") {
+            return formatInteger(value);
+        } else if (this.props.type === "float") {
+            return formatFloat(value);
```
You can restore this diff, I don't see the importance. But in general, I'm okay with this PR.

caburj replied:
Oops. Okay. My bad.

caburj commented on addons/web/static/tests/views/fields/properties_field_tests.js:
```diff
@@ -549,7 +549,9 @@ QUnit.module("Fields", (hooks) => {
             assert.strictEqual(input.value, expected);
         }
 
+        await editValue("0", "0");
         await editValue("2", "2.00");
+        await editValue("0.0", "0");
```
This is for the integer part below. Here, the expected value should be "0.00". After that, the tests will be okay and I'm okay with this PR.

std-odoo commented on addons/web/static/tests/views/fields/properties_field_tests.js:
```diff
@@ -549,6 +549,7 @@ QUnit.module("Fields", (hooks) => {
             assert.strictEqual(input.value, expected);
         }
 
+        await editValue("0", "0.0");
```
this line is failing, it should be `0.00` and not `0.0`

std-odoo commented on addons/web/static/tests/views/fields/properties_field_tests.js:
```diff
@@ -560,6 +561,7 @@ QUnit.module("Fields", (hooks) => {
         await changeType(target, "integer");
         await closePopover(target);
 
+        await editValue("0", "0");
```
can we also add a python test for both `value` and `default` key ?
in `test_properties_field_integer_float_boolean`, write zero and check that we read zero and not false

std-odoo commented on addons/web/static/src/views/fields/properties/property_value.js:
```diff
@@ -200,15 +200,15 @@ export class PropertyValue extends Component {
         } else if (this.props.type === "integer") {
             try {
-                newValue = parseInteger(newValue) || 0;
+                newValue = newValue === "" ? false : parseInteger(newValue);
```
is `newValue === "" ? false` correct ?
with normal field if you empty the <input/> is set to value to 0.00, not to false
See https://drive.google.com/file/d/1b9maXuvASMQutJ7V1t9hwRyg0_0_1jAO/view?usp=sharing

std-odoo commented on addons/web/static/src/views/fields/properties/property_value.js:
```diff
@@ -149,6 +149,10 @@ export class PropertyValue extends Component {
         if (this.props.type === "many2one" && value && value.length === 2) {
             return formatMany2one(value);
+        } else if (this.props.type === "integer") {
+            return typeof value === "undefined" ? "" : formatInteger(value);
```
same here, I wonder if we shouldn't always fallback to zero

std-odoo reviewed (commented):
@julien-banken Hi :) 

Some questions / propositions
Thanks for the fix :)

std-odoo commented on addons/web/static/src/views/fields/properties/property_value.js:
```diff
@@ -149,6 +149,10 @@ export class PropertyValue extends Component {
         if (this.props.type === "many2one" && value && value.length === 2) {
             return formatMany2one(value);
+        } else if (this.props.type === "integer") {
+            return value === undefined || value === false ? "0" : formatInteger(value);
```
can we do something like this ? (same bellow)
```suggestion
            return formatInteger(value || "0");
```

std-odoo commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -734,6 +734,19 @@ def test_properties_field_integer_float_boolean(self):
             self.message_1.attributes[2]['value'], False,
             msg='Boolean value must have been converted to False')
 
+        self.message_1.attributes = [{
+            'name': 'int_value',
+            'value': 0
+        }, {
+            'name': 'float_value',
+            'value': 0
+        }]
+
```
everything can be one lined here

std-odoo commented on odoo/fields.py:
```diff
@@ -3706,10 +3706,12 @@ def _list_to_dict(cls, values_list):
         dict_value = {}
         for property_definition in values_list:
-            property_value = property_definition.get('value') or False
+            property_value = property_definition.get('value', False)
```
is this change still relevant ?

std-odoo commented on odoo/fields.py:
```diff
@@ -3706,10 +3706,12 @@ def _list_to_dict(cls, values_list):
             property_model = property_definition.get('comodel')
 
+            if property_type in ('integer', 'float') and property_value is False:
+                property_value = 0
```
we can change the next line "if -> elif"

std-odoo commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -734,6 +734,19 @@ def test_properties_field_integer_float_boolean(self):
             self.message_1.attributes[2]['value'], False,
             msg='Boolean value must have been converted to False')
 
+        self.message_1.attributes = [{
```
can you also test the behavior on the "default" key on the parent definition ?

std-odoo reviewed (commented):
@julien-banken Hi :) 

Some last polyshing comments and we should be good

Thanks for making properties better :)

std-odoo reviewed (commented):
@julien-banken Hi :) 

I think you try to support a format not supported by the properties, that's why you need to do all this stuff in `_add_default_values`

(some tests do it for simplicity, but in reality they shouldn't and the web client will never send that format)

Cheers

julien-banken commented:
Hello @std-odoo :wave:,

Thank you for the good follow up :smile: I updated the code as you suggested and I think I start to understand the hypothesis you made for the property fields.

Just to recap the current impact of the changes:

On the server side:
- When no value is provided, we will simply store the value `False` as it is currently the case.
- When the user sets the value `0`, we should store `0` and not `False`. I added two tests for that: One to check that the value `0` is not transformed to `False` and one to check that the system will not fallback to the default value when the user sets the value `0`. 

On the client side: 
- When the value of the property fields is `False`, we will simply set the input field value to `0` or `0.00`.
- When the value is invalid, we will fallback to `0` and `0.00`. This is not exactly the same behavior as the standard fields but I think it should be okay for now. Adding that behavior required more important changes and may not enter in the scope of a fix.

std-odoo reviewed (approved):
@julien-banken Nice :) Change is small in the end in fields.py which is a good thing

Thanks for your work

std-odoo commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -734,6 +734,54 @@ def test_properties_field_integer_float_boolean(self):
+        # 0 to False (-> unset value).
+
+        self.message_1.attributes = {'int_value': 0, 'float_value': 0}
+        self.assertEqual(len(self.message_1.attributes), 3)
```
note that we can use _get_sql_properties to even check the value stored in database

std-odoo commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -734,6 +734,54 @@ def test_properties_field_integer_float_boolean(self):
+        self.assertEqual(self.message_1.attributes[0]['value'], 0)
+        self.assertEqual(self.message_1.attributes[1]['value'], 0)
+        self.assertEqual(self.message_1.attributes[2]['value'], False)
+        self.assertEqual(type(self.message_1.attributes[0]['value']), type(0))
+        self.assertEqual(type(self.message_1.attributes[1]['value']), type(0))
+        self.assertEqual(type(self.message_1.attributes[2]['value']), type(False))
```
why this ? (generally we do `isinstance(x, bool)`

std-odoo commented on odoo/addons/test_new_api/tests/test_properties.py:
```diff
@@ -734,6 +734,54 @@ def test_properties_field_integer_float_boolean(self):
+            'author': self.user.id,
+            'attributes': {
+                'int_value': 0,
+                'float_value': 0
```
picky std :p (same above)
```suggestion
                'float_value': 0,
```

(but in that case it can be inlined to be consistent with code below)

std-odoo reviewed (commented):
@julien-banken Some picky annoying remarks :)
